ext/uri/php_uriparser.h

```diff
@@ -0,0 +1,39 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Máté Kocsis <kocsismate@php.net>                            |
+   +----------------------------------------------------------------------+
+*/
+
+#ifndef PHP_URIPARSER_H
+#define PHP_URIPARSER_H
+
+#include <uriparser/Uri.h>
+#include "uriparser/src/UriMemory.h"
+#include <php_uri_common.h>
+
+extern const uri_handler_t uriparser_uri_handler;
+
+typedef struct uriparser_uris_t {
+	UriUriA *uri;
+	zend_string *uri_str;
+	UriUriA *normalized_uri;
+	zend_string *normalized_uri_str;
+} uriparser_uris_t;
```

Comment:
```suggestion
typedef struct uriparser_uris_t {
	UriUriA uri;
	zend_string *uri_str;
	UriUriA normalized_uri;
	zend_string *normalized_uri_str;
} uriparser_uris_t;
```

From what I see, the uriparser library does not allocate `UriUriA` structs itself, but requires you to pass in an OUT-pointer. Thus we can just stack-allocate it / embed it into another struct.

Reply:
The `uri` field is a good candidate indeed, but the `normalized_uri` field is initialized on demand (by `uriparser_read_uri`). So we would have to track its initialization status by a separate bool field because `UriUriA` doesn't provide any indication. Is it still desirable to allocate on stack even if it causes some maintainability degradation?

Reply:
> Is it still desirable to allocate on stack even if it causes some maintainability degradation?

I feel that a separate bool would also be simpler on the maintenance, since you would have less allocs / frees to deal with, but use whatever “feels comfortable”.